### PR TITLE
Tool Python - Run GDScript Toolkit

### PR DESCRIPTION
**Pull request by Numenter**
# Description
Added a python script for simpler use of the GDScript Toolkit. #52

## Type of change
- [x] Tooling (non breaking change that adds functionality to the workflow)

## Output
image

### CODE REVIEW

File: tools/run_gd_script_toolkit.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+if sys.version_info < (3, 4):
+    print('Please upgrade your Python version to 3.4.0 or higher')
+    sys.exit()
+
+game_path = "../game/"
+try:
+    re = subprocess.run(["gdformat", str(game_path)], capture_output=True)
+    print("--- Format ---")
+    print(re.stdout)
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: You should decode the bytes here to string before printing

File: tools/run_gd_script_toolkit.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    print()
+    re = subprocess.run(["gdlint", str(game_path)], capture_output=True)
+    print("--- Lint ---")
+    print(re.stdout)
+except OSError:
+    wait = input("ERROR: GDScript Toolkit not installed!")
+    sys.exit()
+
+print()
+wait = input("Press Enter to close.")
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: Why would you wait here?

File: tools/run_gd_script_toolkit.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+try:
+    re = subprocess.run(["gdformat", str(game_path)], capture_output=True)
+    print("--- Format ---")
+    print(re.stdout)
+    print()
+    re = subprocess.run(["gdlint", str(game_path)], capture_output=True)
+    print("--- Lint ---")
+    print(re.stdout)
+except OSError:
+    wait = input("ERROR: GDScript Toolkit not installed!")
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: Why would you wait here?

File: tools/run_gd_script_toolkit.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+game_path = "../game/"
+try:
+    re = subprocess.run(["gdformat", str(game_path)], capture_output=True)
+    print("--- Format ---")
+    print(re.stdout)
+    print()
+    re = subprocess.run(["gdlint", str(game_path)], capture_output=True)
+    print("--- Lint ---")
+    print(re.stdout)
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: You should decode the bytes here to string before printing

File: tools/run_gd_script_toolkit.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import sys
+import subprocess
+
+if sys.version_info < (3, 4):
+    print('Please upgrade your Python version to 3.4.0 or higher')
+    sys.exit()
+
+game_path = "../game/"
+try:
+    re = subprocess.run(["gdformat", str(game_path)], capture_output=True)
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: why do you try to convert the game_path string to a string?

File: tools/run_gd_script_toolkit.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    print("--- Format ---")
+    print(re.stdout)
+    print()
+    re = subprocess.run(["gdlint", str(game_path)], capture_output=True)
```
Comment (Joshix-1) [inline tools/run_gd_script_toolkit.py]: why do you try to convert the game_path string to a string?

**Review by Joshix-1 — CHANGES_REQUESTED** [review]
I would put all the code in a main function and only run it
`if __name__ == "__main__":` that's a best practise I would follow
You should also try to exit with a non 0 code if linting failed or something hat to be formatted. 
You could let the main function return the exit code and then exit with the code returned by main.
Also add an option to only suggest the formatting and not do it. Then we could run this script in the workflow.

Also please add a Shebang to the top of the script for linux users and mark it executable (chmod +x )

**Comment by BjoernAkAManf** [comment]
Just running subprocesses also seems hacky.
Still questioning the usefulness, but at least it is cross platform.

**Comment by Joshix-1** [comment]
I would do it something like the following: https://github.com/Joshix-1/Suffragium/commit/3d34685d0178afb7b4596dc958cb1eb6d6bab8d3
(That is linted with pylint, formatted with black)

File: tools/run_gd_script_toolkit.py (hunk 2 above)
Comment (Numenter) [reply tools/run_gd_script_toolkit.py]: Windows closes the cmd window after the script is over.

**Comment by Joshix-1** [comment]
Found few issues https://github.com/Joshix-1/Suffragium/commit/cb8d46cf01118345f8b8e0a0b4ad834f6cd719ac
The game_path didn't work on linux

And I marked it as executable with chmod +x

**Comment by Numenter** [comment]
game_path was in windows style. I use normpath to fix it, dont know if this works. (I don't use linux)

**Comment by Joshix-1** [comment]
That doesn't work, my solution worked, and should also work on windows
With your solution I don't get any output from the lint that tells me it didn't worked

Also please mark the file executable

**Comment by Joshix-1** [comment]
the mode of the file didn't change
Also printing the error code doesn't make any sense if it is a number returned from subprocess.run
And it will get printed twice if it is a string. One time with sys.exit and the other time with the print

**Comment by Numenter** [comment]
I change the chmod with "git update-index --chmod=+x run_gdtoolkit.py"
and "git ls-files -s run_gdtoolkit.py" shows that it has 755, but when i commit its back to 644.

**Comment by Joshix-1** [comment]
`chmod +x`: https://github.com/Joshix-1/Suffragium/commit/417a461a03b0f05e878fbd0345bc2a28bd5def5b

**Comment by Numenter** [comment]
I thing I got it.

> `chmod +x`: [Joshix-1@417a461](https://github.com/Joshix-1/Suffragium/commit/417a461a03b0f05e878fbd0345bc2a28bd5def5b)

Thangs for that, but I need to get it myself :P

**Comment by Joshix-1** [comment]
I would format it with black (which makes pylint happy too): https://github.com/Joshix-1/Suffragium/commit/865366cf3db5ed68a2fd016b2b4d69c4de87f83b

We could also use this script in the lint workflow

**Comment by Numenter** [comment]
So, we're gonna comply with PEP 8 (Tabs vs Spaces).

**Comment by Numenter** [comment]
> Just running subprocesses also seems hacky.

How not to use subprocess, enlighten me. @BjoernAkAManf 
Best I can to is get paths to the scripts and now?
```
import gdtoolkit
gdtoolkit_path = os.path.dirname(gdtoolkit.__file__)
gdformat_path = os.path.join(gdtoolkit_path, "formatter")
gdlint_path = os.path.join(gdtoolkit_path, "linter")
```
Import gdformat and import gdlint are not working for me.

**Comment by BjoernAkAManf** [comment]
Cant look myself right now, but using import for path resolution is better than relying on shell / manual path parsing.

Id have a look at their respective main files though. Not that good with python though.

**Comment by Joshix-1** [comment]
gdlint uses `from gdtoolkit.linter.__main__ import main` and gdformat uses `from gdtoolkit.formatter.__main__ import main`
These both don't look like they are stable API. Both rely on sys.argv and the first even calls sys.exit. It would be really hacky to use those functions
I think just calling the commands (like we do now) is the best solution.

**Comment by Joshix-1** [comment]
With the current solution there are a few problems:
- the stderr is only printed if the scripts didn't print anything to stdout.
- And on windows without --skip err gets printed twice even if it is an int, the user doesn't need to see.
- game_path is wrong when called from within the tools folder with ./run_gdtoolkit.py (fixed by calling abspath, which removes ./ from path)
- the script isn't typed (small problem)

This commit fixes all of that: https://github.com/Joshix-1/Suffragium/commit/bf755487c797ec6a237ab5393e08de03e1f3ec7f

**Review by Joshix-1 — APPROVED** [review]

**Review by Joshix-1 — APPROVED** [review]

**Pull request merged**